test/Test/Core/Repo.hs

```diff
@@ -1 +1,65 @@
-module Test.Core.Repo where
\ No newline at end of file
+{-# LANGUAGE QuasiQuotes #-}
+
+module Test.Core.Repo 
+       ( repoRoundtripProp
+       )where
+
+import IW.App (AppEnv, WithError)
+import IW.Core.Id (Id (..))
+import IW.Core.Repo (Repo (..), RepoName (..), RepoOwner (..))
+import IW.Core.SqlArray (SqlArray (..))
+import IW.Effects.Log (runAppLogIO)
+import IW.Db (WithDb)
+import IW.Db.Functions (asSingleRow, query)
+
+import Database.PostgreSQL.Simple.Types ((:.) (..))
+
+import Hedgehog
+import qualified Hedgehog.Gen as Gen
+import qualified Hedgehog.Range as Range
+
+
+repoViaSql :: (WithDb env m, WithError m) => Repo -> m Repo
+repoViaSql repo = asSingleRow $ query
+    [sql| SELECT ?, ?, ?, ?, (? :: TEXT ARRAY) |]
+    (Only (repoId repo) :. repo)
+
+repoRoundtripProp :: AppEnv -> Property
+repoRoundtripProp env = property $ do
+    generatedRepo <- forAll genRepo
+    parsedRepo <- liftIO $ runAppLogIO env $ repoViaSql generatedRepo 
+    parsedRepo === Right generatedRepo
+
+testCategories :: [Text]
+testCategories = 
+    [ "FFI"
+    , "Text"
+    , "Database"
+    , "JSON"
+    , "Concurrency"
+    ]
+
+genRepo :: MonadGen m => m Repo
+genRepo = do
+    repoId         <- genId
+    repoOwner      <- genOwner
+    repoName       <- genName
+    repoDescr      <- genDescr
+    repoCategories <- genCategories
+
+    pure Repo{..} 
+  where
+    genId :: MonadGen m => m (Id Repo)
+    genId = Id <$> Gen.int (Range.constant 1 500)
+
+    genOwner :: MonadGen m => m RepoOwner
+    genOwner = RepoOwner <$> Gen.text (Range.constant 1 20) Gen.alphaNum
+    
+    genName :: MonadGen m => m RepoName
+    genName = RepoName <$> Gen.text (Range.constant 1 20) Gen.alphaNum
```

Comment:
Let's create `Test.Gen` module and move all common generators for `Repo` and `Issue` there, so we can just import them and reuse.

Reply:
Is there some way I can factor out `genId` too?

Reply:
You can just generate `Id a`. For tests it doesn't matter, what type parameter is there